tests/cpu_bugs/kvm_guest_migration.pm

```diff
@@ -164,6 +171,7 @@ sub run {
                 }
                 else {
                     mutex_unlock "flag_lock";
+                    assert_script_run("echo DEST side wait HOST side.");
```

Comment:
for sure no point to do assert_script_run for `echo` message . and maybe better just replcae it with `record_info` ?

Reply:
Its real purpose is to generate an effective delay for waiting for the peer program. because we don't allow use sleep() like function.

Reply:
Not allowing `sleep()` but then simulating that with a call to `echo` shows me that someone did not understand the real problem. There should always be a reason why something "is not allowed". Understanding the reasons is important. It is pretty easy to find many references explaining why a call to `sleep` can be harmful even though there are valid use cases for `sleep`. See https://lmgtfy.com/?q=why+sleep+is+bad+in+code for example.

You are using `sleep()` or are simulating a sleep-like functionality. This is introducing the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. It is pretty easy to find many references explaining why a call to `sleep` can be harmful even though there are valid use cases for `sleep`. See https://lmgtfy.com/?q=why+sleep+is+bad+in+code for example.

We are waiting because some condition changes some time ... keyword(s) is/are *some time*. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time.

Reply:
> Not allowing `sleep()` but then simulating that with a call to `echo` shows me that someone did not understand the real problem. 
Me? I never complain to "Not allowing `sleep()`".
> There should always be a reason why something "is not allowed". Understanding the reasons is important. It is pretty easy to find many references explaining why a call to `sleep` can be harmful even though there are valid use cases for `sleep`. See https://lmgtfy.com/?q=why+sleep+is+bad+in+code for example.
> 
> You are using `sleep()` or are simulating a sleep-like functionality. 
I think the reason for 'simulating', there are two machines "parallel run testcase", OpenQA doesn't provide an API to sync them by the condition(might be it has, but I don't familiar with them). If there is no "waiting"(unblocking), one of the machines has to loop infinitely. I think it consumes large compute power without any benefits. I originally idea is using "any" shell command to simulating a "schedule()" functionality. so that it could "yield", the peer program could hold the lock at this time window.

>This is introducing the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. It is pretty easy to find many references explaining why a call to `sleep` can be harmful even though there are valid use cases for `sleep`. See https://lmgtfy.com/?q=why+sleep+is+bad+in+code for example.
> 
> We are waiting because some condition changes some time ... keyword(s) is/are _some time_. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time.
Yes, check a shared file in NFS directory, I have it. but it also needs a bash shell to execute, seems like an echo command.  
I believe you should have some experience on "parallel programming", would you please give me a hint for how to sync these test cases. 

- add an extra lock, 

- add condition variable?

- semaphore also my preference, does OpenQA provide this API?

Reply:
os-autoinst offers http://open.qa/docs/#_test_synchronization_and_locking_api with mutexes and barriers. This is used just one line above. So it is used already. As an alternative one could also use the network between the machines, e.g. as you mentioned by looking at a shared file in NFS directory. I do not know why specifically you need to wait longer after the mutex_unlock which should already provide the necessary synchronisation.

Reply:
after mutex_unlock(), code will jump to line 147, mutex_lock 'flag_lock', so the unhold lock time window is very tiny. so that peer program hardy holds the lock, even it is ready.
actually, I don't found more easy way for this. call for help.